{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The dNetwork gene network visualization tool\n",
    "\n",
    "This notebook describes how to use and interpret the network exploration and visualization\n",
    "capabilities of the `jp_gene_viz.dNetwork.NetworkDisplay` IPython widget.\n",
    "This presentation assumes that the system running IPython has the `jp_gene_viz`\n",
    "module and dependancies installed.\n",
    "\n",
    "The `NetworkDisplay` widget is intended to allow systems biologists to\n",
    "interact with gene expression networks.  Technically the gene \n",
    "expression networks presented by a `NetworkDisplay` consist of \n",
    "\n",
    "- named nodes representing genes and \n",
    "- weighted directed edges which connect genes where the source gene regulates the target gene.\n",
    "\n",
    "Edges can represent activation or repression relations between genes.  For\n",
    "example the screenshot below displays portion of a regulatory network\n",
    "showing genes regulated by the\n",
    "`stat3` transcription factor gene.  In this case `stat3` is indicated \n",
    "as tending to activate\n",
    "`pdpn` and so the edge between `stat3` and `pdpn` has a positive weight\n",
    "which by default displays in a shade of green.  By contrast `stat3`\n",
    "is indicated as tending to repress `uba2v1` and so the edge between `stat3`\n",
    "and `uba2v1` has a negative weight which by default displays as a shade of\n",
    "red."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![network screenshot](network.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Widget layout\n",
    "\n",
    "The screenshot shown above is annotated with red text to identify the primary\n",
    "layout areas of the `NetworkDisplay` widget.\n",
    "\n",
    "- In the upper left is the graphical display of the nodes and edges of interest.\n",
    "\n",
    "- Along the left are primary graphical controls for the network.\n",
    "\n",
    "- On the bottom is the threshhold, match and status area.\n",
    "\n",
    "- Hidden beneath the status area is a \"settings\" area which is only displayed when the\n",
    "\"settings\" checkbox is checked.\n",
    "\n",
    "### Basic widget creation\n",
    "\n",
    "To create a `NetworkDisplay` widget in an IPython notebook load the `dNetwork`\n",
    "module and install its javascript support in the notebook context and then\n",
    "load a network with data into a widget.  Below is an example which loads data from a file:\n",
    "\n",
    "```\n",
    "from jp_gene_viz import dNetwork\n",
    "dNetwork.load_javascript_support()\n",
    "N = dNetwork.display_network(\"../examples/network.tsv\")\n",
    "```\n",
    "\n",
    "The `display_network` function is a helper function which implements\n",
    "the common use case for loading a network from a file with a default layout.\n",
    "At the bottom of this notebook document is a working example of loading\n",
    "a network which you can use to experiment with the network widget.\n",
    "\n",
    "In the code above the network is loaded from a \"tab separated value\" file\n",
    "consisting of a header line and remaining lines containing at least 3 columns\n",
    "each identifying a regulator gene, a target gene, and the number indicating\n",
    "the degree of regulatory influence.  Here are the first few lines from\n",
    "the `../examples/network.tsv` file:\n",
    "\n",
    "```\n",
    "regulator\ttarget\tbeta.sign.sum\tbeta.non.zero\tvar.exp.median\tvar.exp.ranksum\tprior\n",
    "Exo1\tPpil5\t50\t50\t0.955222567812806\t62\t0\n",
    "Nfyb\tZnrf1\t-50\t50\t0.931630791091037\t196\t0\n",
    "H1f0\tFads2\t50\t50\t0.900272790746536\t670\t0\n",
    ".....\n",
    "```\n",
    "\n",
    "In this case regulator `nfyb` represses target `znrf1` with relative\n",
    "degree of influence -50.\n",
    "\n",
    "### Widget model\n",
    "\n",
    "The `NetworkDisplay` widget displays all or some edges and\n",
    "nodes of the gene network loaded into the widget.\n",
    "Internally the complete data is stored as the \"data graph\".  The current subset of nodes\n",
    "and edges displayed are stored separately as the \"display graph\".  The \n",
    "data and display graphs\n",
    "also maintain layouts which assign two dimensional locations to the nodes\n",
    "which determines their position on the graph canvas when they are displayed..\n",
    "\n",
    "Many of the\n",
    "interactive controls of the widgets add or remove data from the display\n",
    "graph or adjust the layout, sometimes by getting information from the data graph and \n",
    "default layout.\n",
    "\n",
    "### The edge threshhold\n",
    "\n",
    "The bottom slider next to the \"threshhold\" button determines the lower limit\n",
    "for weights of edges of interest to be displayed, in absolute value.  For example\n",
    "if the threshhold value is 11, then only edges with weights greater than 11 or\n",
    "more negative than -11 will be shown in the display graph.\n",
    "\n",
    "### The status area\n",
    "\n",
    "The **status area** text box is used to display feedback and data about the\n",
    "state of the widget.  For example when the user clicks **list nodes** the list of\n",
    "node names appears in the status area.\n",
    "\n",
    "### The labels checkbox\n",
    "\n",
    "The **labels** checkbox determines whether the names of genes are displayed\n",
    "as text labels in the network display.  Check the check box to show the labels.\n",
    "\n",
    "### Node name matching\n",
    "\n",
    "The **match** button and the text area to the right of the **match**\n",
    "button are used to select specific node names or node names that match\n",
    "specific patterns.\n",
    "\n",
    "**Use case**:  I want to only see the genes `stat3`, `mt2`, and `foxp1`.\n",
    "\n",
    "**Procedure**: Type `stat3 mt2 foxp1` in the text area next to the **match**\n",
    "button and then click the **match** button.  To improve the layout of the resultant\n",
    "graph then click the **layout** button.\n",
    "\n",
    "**Use case**:  I want to see all nodes with names starting `eg...` or `stat...`.\n",
    "\n",
    "**Procedure**: Type `eg* stat*` in the text area next to the **match**\n",
    "button and then click the match button.  To improve the layout of the resultant\n",
    "graph then click the **layout** button.\n",
    "\n",
    "## Data selection buttons\n",
    "\n",
    "The **trim**, **expand**, **regulates**, **targeted by**, **restore**,\n",
    "and **threshhold** buttons modify which edges are displayed as described\n",
    "in the following subsections.\n",
    "\n",
    "### The restore button\n",
    "\n",
    "The **restore** button sets the displayed nodes and edges to all edges\n",
    "of the data graph which satisfy the current threshhold, using the default\n",
    "data layout to determine the node positions.\n",
    "\n",
    "**Use case**: I want to view all edges and nodes from the originally loaded\n",
    "network.\n",
    "\n",
    "**Procedure**: Slide the threshhold slider all the way to the left until the value\n",
    "displays `0` and then click the **restore** button.\n",
    "\n",
    "### The threshhold button\n",
    "\n",
    "The **threshhold** button applies the current threshhold to the edges currently\n",
    "displayed.\n",
    "\n",
    "**Use case**: Of the edges currently displayed I want to only see those with\n",
    "weight greater than 20 in absolute value.\n",
    "\n",
    "**Procedure**:  Slide the threshhold slider to value 20 and then press the\n",
    "**threshhold** button.\n",
    "\n",
    "### The trim button\n",
    "\n",
    "The **trim** button eliminates from display all nodes which do not regulate\n",
    "a currently displayed target.\n",
    "\n",
    "**Use case**:  Of the currently displayed nodes and edges I only want to\n",
    "see the nodes shown as regulators, and the displayed edges between\n",
    "those nodes.\n",
    "\n",
    "**Procedure**: Press the **trim** button.\n",
    "\n",
    "### The expand, regulates, and targetted by buttons\n",
    "\n",
    "The **regulates** button adds to the display nodes that are regulated\n",
    "by the current set of displayed nodes (if they are absent,\n",
    "subject to the threshhold constraint).  Similarly\n",
    "the **targetted by** button adds to the display nodes that regulate some\n",
    "node in the set of currently displayed nodes.  The\n",
    "**expand** button adds nodes that either regulate or are regulated by the\n",
    "currently visible node set.\n",
    "\n",
    "**Use case**:  I want to see the part of the network which shows\n",
    "`stat3` and the genes which directly regulate `stat3` (subject to the\n",
    "threshhold constraint).\n",
    "\n",
    "**Procedure**: Start by selecting only `stat3` by typing `stat3` in the\n",
    "text area next to the **match** button.  Press the **match** button\n",
    "to display `stat3` by itself.  Then add the genes that target\n",
    "`stat3` by clicking the `targeted by` button.  Improve the layout of\n",
    "the network subset by clicking the **layout** button.\n",
    "\n",
    "### The layout button and drop down\n",
    "\n",
    "The **layout** button applies a network layout heuristic to the currently\n",
    "displayed network subset.  A dropdown above the **layout** button selects\n",
    "the heuristic to use for layouts among a number of options.\n",
    "\n",
    "**Use case**: I want to see the genes that regulate `stat3` and the\n",
    "genes that regulate those genes also, layed out nicely.\n",
    "\n",
    "**Procedure**: Type `stat3` in the **match** text area and click the **match**\n",
    "button to display `stat3` by itself.  Click the `targeted by` button twice\n",
    "to get the genes that regulate `stat3` and their regulators.  Then click\n",
    "`layout` to improve the `layout` for the network subset shown.  For an alternate\n",
    "layout switch the dropdown heuristic above the layout to **skeleton**\n",
    "and click **layout** again.\n",
    "\n",
    "### Listing visible nodes and edges\n",
    "\n",
    "The **list nodes** and **list edges** buttons are used to display \n",
    "reports of the nodes and edges of the display graph in the status area.\n",
    "\n",
    "**Use case**:  I want to get a report listing the currently displayed\n",
    "edges with their respective weights.\n",
    "\n",
    "**Procedure**:  Click the **list edges** button and then copy/paste \n",
    "the report of visible edges from the widget status text box area.\n",
    "\n",
    "## Visual selection operations\n",
    "\n",
    "The network widget supports a square selection mechanism which\n",
    "highlights a squere region of the displayed network.  The\n",
    "**zoom**, **focus**, and **ignore** buttons operate on elements\n",
    "identified by the square selection.\n",
    "\n",
    "To create a rectangular selection **shift-click** at one extreme point\n",
    "of the square in the network display area and then **click** at an opposing\n",
    "extreme point.  Then use one of the buttons to perform a selection operation\n",
    "as described below.  To cancel a selection, click again anywhere in the\n",
    "network display area.\n",
    "\n",
    "### The zoom button\n",
    "\n",
    "The **zoom** button zooms the network view to only show the portion\n",
    "of the network under the selection.  Edges that reach\n",
    "to nodes not under the selection be shown drawing off the edge of the\n",
    "display.  The selection will cover the display after zooming -- click\n",
    "anywhere in the display to cancel the selection.\n",
    "\n",
    "### The focus button\n",
    "\n",
    "The **focus** button displays only nodes and edges under the selection,\n",
    "ignoring all others.  For example edges that reach out of the selection\n",
    "will not be shown.\n",
    "\n",
    "### The ignore button\n",
    "\n",
    "The **ignore** button removes nodes and edges under the selection from display.\n",
    "\n",
    "**Use case**:  I want to remove `mxi1` and all its incident edges from the\n",
    "current network display.\n",
    "\n",
    "**Procedure**: Position the selection over `mxi1` and click the **ignore** button.\n",
    "\n",
    "### The draw button\n",
    "\n",
    "The draw button redraws the graph, sometimes recentering the display.\n",
    "For example to \"undo\" the result of clicking the **zoom** button, click\n",
    "the **draw** button.\n",
    "\n",
    "### Manually moving labels\n",
    "\n",
    "When labels are shown they may be moved.  To move a label click the label\n",
    "and then click again where you want to place the label.  Operations that\n",
    "modify graph layout will have the effect of forgetting manually positioned\n",
    "labels.\n",
    "\n",
    "### Manually moving nodes (and descendant nodes...)\n",
    "\n",
    "Nodes may be positioned manually also.  To move a node click on the node and click\n",
    "again where you want the node to be positioned.  It is also possible to\n",
    "move a node and all nodes it directly regulates:  move the depth slider to\n",
    "`1` and then position the node as before.  With the depth at `1` all regulated\n",
    "nodes will move together with the node.  Deeper descendant nodes can be\n",
    "positioned in a similar manner by setting the depth slider to higher values.\n",
    "Operations that\n",
    "modify graph layout will have the effect of forgetting manually positioned\n",
    "nodes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "![settings screenshot](dNetwork_settings.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## The settings controls\n",
    "\n",
    "More advanced controls are in a settings area which is hidden by default.\n",
    "The advanced controls support:\n",
    "\n",
    "- Adjusting the label size for regulators and non-regulators.\n",
    "- Adjusting the node and edge color interpolation scale.\n",
    "- Manual override colorization for individual nodes, edges, and labels.\n",
    "- Saving the network display state to a file on the Jupyter server.\n",
    "- Loading network state from a file on the Jupyter server.\n",
    "- Uploading or downloading files from the Jupyter server to the \"local\" file system.\n",
    "- Snapshotting a network visualization to an image format.\n",
    "\n",
    "Check the settings checkbox to show the settings area.\n",
    "\n",
    "### Adjusting label sizes\n",
    "\n",
    "In the settings area slider marked **labels** adjusts the minimum size for all labels.\n",
    "The settings marked **tf labels** separately adjusts the size for labels that are\n",
    "transcription factors (genes that regulate other genes).  The network will not\n",
    "automatically redraw when you change these sliders -- click the **draw** button to\n",
    "redraw the network after adjusting the sliders as desired.\n",
    "\n",
    "**Use case**: I want to show labels for the currently displayed regulators\n",
    "at font size 12, and I don't want to show labels for non-regulators.\n",
    "\n",
    "**Procedure**:  Make sure the **labels** checkbox is checked so labels are\n",
    "enabled.  Slide the **tf labels** slider to 20 to set the font size for regulators.\n",
    "Slide the **labels** slider to 0 to disable labels for non-regulators.\n",
    "Click the **draw** button to redraw the network display with the new settings.\n",
    "\n",
    "### Adjusting color interpolation scales\n",
    "\n",
    "Below the labels sliders are two color interpolation widgets\n",
    "for adjusting the interpolation of **node** colors and **edge** colors.\n",
    "The widgets have a color selection area above and a color interpolation\n",
    "bar below.  To add breakpoints to the interpolation bar select the\n",
    "desired color from the selection area by clicking the desired color\n",
    "and click again at the desired position in the interpolation bar.\n",
    "To remove a color breakpoint from the interpolation bar, click the\n",
    "circle identifying the breakpoint on the interpolation bar and click\n",
    "again in the color selection area.\n",
    "\n",
    "When the colors are adjusted as desired click the **draw** button\n",
    "to redraw the network display with the new interpolation breakpoints.\n",
    "\n",
    "### Manual colorization\n",
    "\n",
    "The **manual colorize** checkbox in the settings area enables manual colorization\n",
    "mode for nodes, edges and labels.  When the **manual colorize** checkbox is checked\n",
    "other mouse interactions in the network display area will not behave normally,\n",
    "so be sure to uncheck it when you are done adjusting colors manually.\n",
    "Manually colorized objects will appear slightly larger than normal until\n",
    "they are re-drawn using the **draw** button.  The **reset default** button\n",
    "will \"forget\" all manual colorization.\n",
    "\n",
    "**Use case**:  I want to show a visualization of the genes targetted by `egr3`\n",
    "with the `egr3` node and its label both colored purple.\n",
    "\n",
    "**Procedure**:  Proceed as follows:\n",
    "\n",
    "- Make sure the labels checkbox is checked.\n",
    "- Type `egr3` in the match area and click the **match** button to select `egr3` by itself.\n",
    "- Click the **regulates** button to add the nodes regulated by `egr3` to the display.\n",
    "- Click the **layout** button to improve the layout of the selected nodes.\n",
    "- Possibly adjust the positions of the nodes and labels by mouse interactions.\n",
    "- Click the **settings** checkbox and the and the **manual colorize** checkbox in the settings area.\n",
    "- In the manual colorize color selection area click the purple square.\n",
    "- In the network area click the `egr3` node and label to change them to purple.\n",
    "- Uncheck the **manual colorize** checkbox to return to normal mouse interaction mode.\n",
    "- Click the **draw** button to display the colorized object at normal size.\n",
    "- [Take a snapshot of the visualization as described below.]\n",
    "\n",
    "### Saving network display state in the Jupyter server file system\n",
    "\n",
    "The **save** button will save the network in an archive format which can be restored\n",
    "at a later time.  The saved file will be on the \"Jupyter server file system\" which\n",
    "may be the same as your local file system if you are running Jupyter locally.  If you are not\n",
    "running locally you may wish to **save** the file to the Jupyter server file system\n",
    "and then upload the file to your local file system as described below.\n",
    "\n",
    "The **save** button requires an appropriate file name or path in the text area\n",
    "next to the **save/load file name** button.\n",
    "\n",
    "**Use case**: I want to archive the current visualization state to `viz.zip` on the\n",
    "Jupyter server file system.\n",
    "\n",
    "**Procedure**: Type `viz.zip` in the text area next to the save/load file name button\n",
    "and click the **save** button.  An alert will confirm that the state is saved to the file.\n",
    "\n",
    "### Restoring network display state from the Jupyter server file system\n",
    "\n",
    "The **load** button will load a network from an archive format saved in a file\n",
    "in the \"Jupyter server file system\".  If the file you wish to restore is not presently \n",
    "in the server file system you may upload the file using the **upload/download** button\n",
    "as described below.\n",
    "\n",
    "**Use case**: I want to restore the network state from an archive on the server\n",
    "file system, but I don't remember the file name of the archive.\n",
    "\n",
    "**Procedure**: Click the **save/load filename** button to open a dialog which displays\n",
    "the files on the server file system.  Click the file you saved previously once you identify\n",
    "it (let's say it is `viz.zip`).  Close the dialog.  A file path ending `viz.zip` should\n",
    "now show in the **save/load file name** text area.  Click the **load** button to load\n",
    "the archive.  If there are no errors an alert will confirm that the state has been loaded.\n",
    "\n",
    "### Uploading/downloading files to and from the local file system and the Jupyter server file system\n",
    "\n",
    "The **upload/download** button opens a dialog which allows uploading files from\n",
    "the local file system to the Jupyter server file system and also allows downloading\n",
    "files from the Jupyter server file system to the local file system.\n",
    "\n",
    "**Use case**:  I want to download the archive `viz.zip` which I stored previously on\n",
    "the server file system to my local `Downloads` folder.\n",
    "\n",
    "**Procedure**:  Click the **upload/download** button.\n",
    "In the upload/download dialog lower area, click the `viz.zip` file name.\n",
    "The browser should automatically download the file.\n",
    "\n",
    "**Use case**: I want to upload the archive `viz0.zip` from my local `Downloads` folder\n",
    "to the remote server file system.\n",
    "\n",
    "**Procedure**:  Click the **upload/download** button.\n",
    "In the upload/download dialog upper area click the **Choose file** button\n",
    "next to the upload label.  A local file system dialog will open to allow\n",
    "you to choose the local file.  When you select the `viz0.zip` from the\n",
    "local file system the upload/download dialog will report \"Uploaded 'viz0.zip'\".\n",
    "\n",
    "### Snapshotting a visualization as an image\n",
    "\n",
    "The **snapshot** button creates an image version of the current network\n",
    "visualization.\n",
    "\n",
    "**Use case**: I want to download an image version of the current network\n",
    "visualization with 600 pixels per side, using the filename `Figure3.png`\n",
    "\n",
    "**Procedure**:  Adjust the **side** slider next to the **snapshot** button\n",
    "until the value reads `600`.  Change the text in the **snapshot filename**\n",
    "text area to read `Figure3`.\n",
    "Click the **snapshot** button.  The browser\n",
    "should automatically download a `600x600` PNG format as `Figure3.png`.\n",
    "\n",
    "**Note**: For large and complex images the method used to download\n",
    "the images may silently fail.  If so, another alternative is to take\n",
    "a screen shot of the browser presentation to get an image of the\n",
    "visualization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "require([\"widgets/js/widget\", \"widgets/js/manager\"], function(widget, manager){\n",
       "    \n",
       "    var svgEventHandlerFactory = function(that) {\n",
       "        var svgEventHandler = function(e) {\n",
       "            //debugger;\n",
       "            var target = e.target;\n",
       "            var info = {};\n",
       "            for (var attr in e) {\n",
       "                var val = e[attr];\n",
       "                var ty = (typeof val);\n",
       "                if ((ty == \"number\") ||\n",
       "                    (ty == \"string\") ||\n",
       "                    (ty == \"boolean\")) {\n",
       "                    info[attr] = val;\n",
       "                }\n",
       "            }\n",
       "            info.name = target.ipy_name;\n",
       "            var ept = SVGEventLocation(that, e);\n",
       "            info.svgX = ept.x;\n",
       "            info.svgY = ept.y;\n",
       "            var json = JSON.stringify(info);\n",
       "            that.model.set(\"event\", json);\n",
       "            that.touch();\n",
       "        };\n",
       "        return svgEventHandler;\n",
       "    };\n",
       "\n",
       "    var SVGEventLocation = function(that, e) {\n",
       "        // http://stackoverflow.com/questions/10298658/mouse-position-inside-autoscaled-svg\n",
       "        var pt = that.reference_point;\n",
       "        var svg = that.$svg[0];\n",
       "        pt.x = e.clientX;\n",
       "        pt.y = e.clientY;\n",
       "        return pt.matrixTransform(svg.getScreenCTM().inverse());\n",
       "    }\n",
       "    \n",
       "    var SVGCanvasView = widget.DOMWidgetView.extend({\n",
       "        \n",
       "        render: function() {\n",
       "            var that = this;\n",
       "            var svg = that.svg_elt(\"svg\");\n",
       "            var eventHandler = svgEventHandlerFactory(that);\n",
       "            that.eventHandler = eventHandler;\n",
       "            that.named_elements = {};\n",
       "            svg.ipy_name = \"\";\n",
       "            that.$svg = $(svg);\n",
       "            that.reference_point = svg.createSVGPoint();\n",
       "            svg.setAttribute(\"preserveAspectRatio\", \"none\");\n",
       "            that.$el.append(that.$svg);\n",
       "            that.svg_parameters_changed();\n",
       "            that.commands_changed();\n",
       "            that.start_watch_event();\n",
       "            that.model.on(\"change:commands\", that.commands_changed, that);\n",
       "            that.model.on(\"change:viewBox\", that.svg_parameters_changed, that);\n",
       "            that.model.on(\"change:width\", that.svg_parameters_changed, that);\n",
       "            that.model.on(\"change:height\", that.svg_parameters_changed, that);\n",
       "            that.model.on(\"change:style\", that.svg_parameters_changed, that);\n",
       "            that.model.on(\"change:watch_event\", that.start_watch_event, that);\n",
       "            that.model.on(\"change:unwatch_event\", that.stop_watch_event, that);\n",
       "        },\n",
       "        \n",
       "        start_watch_event: function() {\n",
       "            //debugger;\n",
       "            var that = this;\n",
       "            var event_types = that.model.get(\"watch_event\");\n",
       "            if (event_types != \"\") {\n",
       "                that.$svg.on(event_types, that.eventHandler);\n",
       "                that.model.set(\"watch_event\", \"\");\n",
       "                that.touch();\n",
       "            }\n",
       "        },\n",
       "        \n",
       "        stop_watch_event: function() {\n",
       "            //debugger;\n",
       "            var that = this;\n",
       "            var event_types = that.model.get(\"unwatch_event\");\n",
       "            if (event_types != \"\") {\n",
       "                that.$svg.off(event_types);\n",
       "                that.model.set(\"watch_event\", \"\");\n",
       "                that.touch();\n",
       "            }\n",
       "        },\n",
       "        \n",
       "        commands_changed: function() {\n",
       "            var that = this;\n",
       "            var svg = that.$svg[0];\n",
       "            var commands = that.get_JSON(\"commands\")\n",
       "            for (var i=0; i<commands.length; i++) {\n",
       "                var command_dict = commands[i];\n",
       "                var indicator = command_dict[\"command\"];\n",
       "                var method = that[\"do_\"+indicator];\n",
       "                method(that, command_dict);\n",
       "            }\n",
       "        },\n",
       "\n",
       "        do_fit: function(that, info) {\n",
       "            // fit viewport to bounding box.\n",
       "            var svg = that.$svg[0];\n",
       "            var bbox = svg.getBBox();\n",
       "            var D = {\"width\": bbox.width, \"height\": bbox.height, \"x\": bbox.x, \"y\": bbox.y}\n",
       "            var vbox = \"\" + D.x + \" \" + D.y + \" \" + D.width + \" \" + D.height;\n",
       "            if ((D.width > 0) && (D.height > 0)) {\n",
       "                that.model.set(\"boundingBox\", D);\n",
       "                if (info.changeView) {\n",
       "                    that.model.set(\"viewBox\", vbox);\n",
       "                }\n",
       "                // Element viewBox will be updated later by model change.\n",
       "                that.touch();\n",
       "            }\n",
       "        },\n",
       "        \n",
       "        do_add_element: function (that, info) {\n",
       "            //debugger;\n",
       "            var tag = info.tag;\n",
       "            var name = info.name;\n",
       "            var element = that.svg_elt(tag);\n",
       "            element.ipy_name = name;\n",
       "            var $element = $(element);\n",
       "            that.update_element($element, info);\n",
       "            // add event callbacks\n",
       "            that.$svg.append($element);\n",
       "            that.named_elements[name] = $element;\n",
       "        },\n",
       "        \n",
       "        do_change_element: function (that, info) {\n",
       "            var name = info.name;\n",
       "            var $element = that.named_elements[name];\n",
       "            if ($element) {\n",
       "                that.update_element($element, info);\n",
       "            } else {\n",
       "                console.warn(\"couldn't find element for \"+name);\n",
       "            }\n",
       "        },\n",
       "        \n",
       "        do_delete: function (that, info) {\n",
       "            var names = info.names;\n",
       "            for (var i=0; i<names.length; i++) {\n",
       "                var name = names[i];\n",
       "                var $element = that.named_elements[name];\n",
       "                if ($element) {\n",
       "                    $element.remove();\n",
       "                    delete that.named_elements[name];\n",
       "                }\n",
       "            }\n",
       "        },\n",
       "        \n",
       "        update_element: function($element, info) {\n",
       "            var element = $element[0];\n",
       "            var atts = info.atts;\n",
       "            var style = info.style;\n",
       "            var text = info.text;\n",
       "            if (atts) {\n",
       "                for (var att in atts) {\n",
       "                    element.setAttribute(att, atts[att]);\n",
       "                }\n",
       "            }\n",
       "            if (style) {\n",
       "                for (var styling in style) {\n",
       "                    element.style[styling] = style[styling];\n",
       "                }\n",
       "            }\n",
       "            if (text) {\n",
       "                $element.empty();\n",
       "                var node = document.createTextNode(text);\n",
       "                element.appendChild(node);\n",
       "            }\n",
       "        },\n",
       "        \n",
       "        do_empty: function (that, info) {\n",
       "            that.named_elements = {};\n",
       "            that.$svg.empty();\n",
       "        },\n",
       "        \n",
       "        svg_parameters_changed: function() {\n",
       "            var that = this;\n",
       "            var style_additions = that.get_JSON(\"style\");\n",
       "            var svg = that.$svg[0];\n",
       "            //debugger;\n",
       "            svg.setAttribute(\"viewBox\", that.model.get(\"viewBox\"));\n",
       "            svg.setAttribute(\"width\", that.model.get(\"width\"));\n",
       "            svg.setAttribute(\"height\", that.model.get(\"height\"));\n",
       "            for (var style_attr in style_additions) {\n",
       "                svg.style[style_attr] = style_additions[style_attr];\n",
       "            }\n",
       "        },\n",
       "        \n",
       "        get_JSON: function(name) {\n",
       "            var json = this.model.get(name);\n",
       "            return $.parseJSON(json);\n",
       "        },\n",
       "        \n",
       "        svg_elt: function(kind) {\n",
       "            return document.createElementNS('http://www.w3.org/2000/svg', kind);\n",
       "        }\n",
       "        \n",
       "    });\n",
       "    \n",
       "    manager.WidgetManager.register_widget_view('SVGCanvasView', SVGCanvasView);\n",
       "});\n"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "// This is the javascript side to match js_proxy.py.\n",
       "// See js_proxy.py for an explanation of the command transfer\n",
       "// protocol.\n",
       "\n",
       "require([\"widgets/js/widget\", \"widgets/js/manager\", \"underscore\", \"jquery\"\n",
       "], function(widget, manager, _, $) {\n",
       "    //debugger;\n",
       "\n",
       "    var JSProxyView = widget.DOMWidgetView.extend({\n",
       "\n",
       "        render: function() {\n",
       "            var that = this;\n",
       "            that.on(\"displayed\", function() {\n",
       "                that.update();\n",
       "            });\n",
       "            // \"new\" keyword emulation\n",
       "            // http://stackoverflow.com/questions/17342497/dynamically-control-arguments-while-creating-objects-in-javascript\n",
       "            that.$el.New = function(klass, args) {\n",
       "                var obj = Object.create(klass.prototype);\n",
       "                return klass.apply(obj, args) || obj;\n",
       "            }\n",
       "\n",
       "            // fix key bindings for wayward element.\n",
       "            // XXXX This is a bit of a hack that may not be needed in future\n",
       "            // Jupyter releases.\n",
       "            that.$el.Fix = function(element) {\n",
       "                debugger;\n",
       "                that.model.widget_manager.keyboard_manager.register_events(element);\n",
       "            }\n",
       "        },\n",
       "\n",
       "        update: function(options) {\n",
       "            var that = this;\n",
       "            var commands = that.model.get(\"commands\");\n",
       "            if (commands.length >= 2) {\n",
       "                var command_counter = commands[0];\n",
       "                var command_list = commands[1];\n",
       "                var level = commands[2];\n",
       "                level = that.check_level(level);\n",
       "                var results = [];\n",
       "                _.each(command_list, function(command,i) {\n",
       "                    var result = that.execute_command(command);\n",
       "                    results[i] = that.json_safe(result, level);\n",
       "                });\n",
       "                that.model.set(\"commands\", []);\n",
       "                that.model.set(\"results\", [command_counter, results])\n",
       "                that.touch();\n",
       "            }\n",
       "        },\n",
       "\n",
       "        execute_command: function(command) {\n",
       "            var that = this;\n",
       "            var result = command;\n",
       "            if ($.isArray(command)) {\n",
       "                var indicator = command[0];\n",
       "                var remainder = command.slice();\n",
       "                remainder.shift();\n",
       "                if (indicator == \"element\") {\n",
       "                    result = that.$el;\n",
       "                } else if (indicator == \"window\") {\n",
       "                    result = window;\n",
       "                } else if (indicator == \"method\") {\n",
       "                    var target_desc = remainder.shift();\n",
       "                    var target = that.execute_command(target_desc);\n",
       "                    var name = remainder.shift();\n",
       "                    var args = remainder.map(that.execute_command, that);\n",
       "                    var method = target[name];\n",
       "                    if (method) {\n",
       "                        result = method.apply(target, args);\n",
       "                    } else {\n",
       "                        result = \"In \" + target + \" no such method \" + name;\n",
       "                    }\n",
       "                } else if (indicator == \"function\") {\n",
       "                    var function_desc = remainder.shift();\n",
       "                    var function_value = that.execute_command(function_desc);\n",
       "                    var args = remainder.map(that.execute_command, that);\n",
       "                    // Use \"that\" as the \"this\" value for function values?\n",
       "                    result = function_value.apply(that, args);\n",
       "                } else if (indicator == \"id\") {\n",
       "                    result = remainder[0];\n",
       "                } else if (indicator == \"list\") {\n",
       "                    result = remainder.map(that.execute_command, that);\n",
       "                } else if (indicator == \"dict\") {\n",
       "                    result = {}\n",
       "                    var desc = remainder[0];\n",
       "                    for (var key in desc) {\n",
       "                        var key_desc = desc[key];\n",
       "                        var val = that.execute_command(key_desc);\n",
       "                        result[key] = val;\n",
       "                    }\n",
       "                } else if (indicator == \"callback\") {\n",
       "                    var identifier = remainder.shift();\n",
       "                    var data = remainder.shift();\n",
       "                    var level = remainder.shift();\n",
       "                    // sanity check\n",
       "                    level = that.check_level(level);\n",
       "                    result = that.callback_factory(identifier, data, level);\n",
       "                } else if (indicator == \"get\") {\n",
       "                    var target_desc = remainder.shift();\n",
       "                    var target = that.execute_command(target_desc);\n",
       "                    var name = remainder.shift();\n",
       "                    result = target[name];\n",
       "                } else if (indicator == \"set\") {\n",
       "                    var target_desc = remainder.shift();\n",
       "                    var target = that.execute_command(target_desc);\n",
       "                    var name = remainder.shift();\n",
       "                    var value_desc = remainder.shift()\n",
       "                    var value = that.execute_command(value_desc);\n",
       "                    target[name] = value;\n",
       "                    result = target;\n",
       "                } else if (indicator == \"null\") {\n",
       "                    target_desc = remainder.shift();\n",
       "                    that.execute_command(target_desc);\n",
       "                    result = null;\n",
       "                } else {\n",
       "                    result = \"Unknown indicator \" + indicator;\n",
       "                }\n",
       "            }\n",
       "            return result;\n",
       "        },\n",
       "\n",
       "        check_level: function(level) {\n",
       "            if ((typeof level) != \"number\" || (level < 0)) {\n",
       "                level = 0;\n",
       "            } else if (level > 5) {\n",
       "                level = 5;\n",
       "            }\n",
       "            return level;\n",
       "        },\n",
       "\n",
       "        callback_factory: function(identifier, data, level) {\n",
       "            var that = this;\n",
       "            // Counter makes sure change is noticed even if other arguments don't change.\n",
       "            var counter = 0;\n",
       "            var handler = function () {\n",
       "                counter += 1;\n",
       "                var payload = that.json_safe([identifier, data, arguments, counter], level + 1);\n",
       "                that.model.set(\"callback_results\", payload);\n",
       "                that.touch();\n",
       "            };\n",
       "            return handler;\n",
       "        },\n",
       "\n",
       "        json_safe: function(val, depth) {\n",
       "            // maybe expand later as need arises\n",
       "            var that = this;\n",
       "            var ty = (typeof val);\n",
       "            if ((ty == \"number\") || (ty == \"string\") || (ty == \"boolean\")) {\n",
       "                return val;\n",
       "            }\n",
       "            if (!val) {\n",
       "                // translate all other falsies to None\n",
       "                return null;\n",
       "            }\n",
       "            if (((typeof depth) == \"number\") && (depth > 0)) {\n",
       "                if ($.isArray(val)) {\n",
       "                    var result = [];\n",
       "                    _.each(val, function(elt, i) {\n",
       "                        var r = that.json_safe(elt, depth-1);\n",
       "                        if (r != null) {\n",
       "                            result[i] = r;\n",
       "                        }\n",
       "                    });\n",
       "                    return result;\n",
       "                } else {\n",
       "                    var result = {};\n",
       "                    for (var key in val) {\n",
       "                        var jv = that.json_safe(val[key], depth-1);\n",
       "                        if (jv != null) {\n",
       "                            result[key] = jv;\n",
       "                        }\n",
       "                    }\n",
       "                    return result;\n",
       "                }\n",
       "            }\n",
       "            return null;\n",
       "        }\n",
       "\n",
       "    });\n",
       "\n",
       "    manager.WidgetManager.register_widget_view('JSProxyView', JSProxyView);\n",
       "});\n"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "/*\n",
       "jQuery plugin for a color arrow cursor over an element.\n",
       "\n",
       "Based on\n",
       "http://stackoverflow.com/questions/18779354/assign-color-to-mouse-cursor-using-css\n",
       "*/\n",
       "\n",
       "(function ($) {\n",
       "\n",
       "    $.fn.color_cursor = function(color) {\n",
       "        var that = this;\n",
       "        var cvs = document.createElement(\"canvas\");\n",
       "        cvs.height = 16;\n",
       "        cvs.width = 16;\n",
       "        var ctx = cvs.getContext(\"2d\");\n",
       "        ctx.strokeStyle = color;\n",
       "        ctx.lineWidth = 4;\n",
       "        ctx.lineCap = \"round\";\n",
       "        ctx.moveTo(2, 12);\n",
       "        ctx.lineTo(2, 2);\n",
       "        ctx.lineTo(12, 2);\n",
       "        ctx.moveTo(2, 2);\n",
       "        ctx.lineTo(30, 30)    \n",
       "        ctx.stroke();\n",
       "        var url = cvs.toDataURL();\n",
       "        that.css(\"cursor\", \"url(\" + url + '), auto');\n",
       "        return that;\n",
       "    };\n",
       "\n",
       "    $.fn.color_cursor_reset = function () {\n",
       "        this.css(\"cursor\", \"\");\n",
       "        return this;\n",
       "    };\n",
       "\n",
       "    $.fn.color_cursor.example = function (element) {\n",
       "        debugger;\n",
       "        var d = $(\"<div>Cursor should be red over me. Click to reset.</div>\")\n",
       "            .color_cursor(\"red\");\n",
       "        var reset = function() {\n",
       "            d.color_cursor_reset();\n",
       "            d.html(\"Now the cursor should be normal\");\n",
       "        }\n",
       "        d.click(reset);\n",
       "        element.append(d);\n",
       "        return element;\n",
       "    };\n",
       "\n",
       "})(jQuery)\n"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Reading network', '../examples/network.tsv')\n",
      "('Loading saved layout', '../examples/network.tsv.layout.json')\n"
     ]
    }
   ],
   "source": [
    "# Here is an example live usage of the widget:\n",
    "\n",
    "# Uncomment to run without install.\n",
    "import sys\n",
    "if \"..\" not in sys.path:\n",
    "    sys.path.append(\"..\")\n",
    "from jp_gene_viz import dNetwork\n",
    "dNetwork.load_javascript_support()\n",
    "N = dNetwork.display_network(\"../examples/network.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
